rules-java/api/src/main/java/org/jboss/windup/rules/apps/java/reporting/rules/CreateDependencyGraphDataRuleProvider.java

```diff
@@ -0,0 +1,158 @@
+import org.jboss.windup.rules.apps.java.condition.SourceMode;
+import org.jboss.windup.rules.apps.java.dependencyreport.DependenciesReportModel;
+import org.jboss.windup.rules.apps.java.dependencyreport.DependencyReportDependencyGroupModel;
+import org.jboss.windup.rules.apps.java.reporting.freemarker.dto.DependencyGraphItem;
+import org.jboss.windup.rules.apps.java.reporting.freemarker.dto.DependencyGraphItemsAndRelations;
+import org.jboss.windup.rules.apps.java.reporting.freemarker.dto.DependencyGraphRelation;
+import org.jboss.windup.util.exception.WindupException;
+import org.ocpsoft.rewrite.config.Configuration;
+import org.ocpsoft.rewrite.config.ConfigurationBuilder;
+import org.ocpsoft.rewrite.context.EvaluationContext;
+
+import java.io.FileWriter;
+import java.io.IOException;
+import java.nio.file.Path;
+import java.util.ArrayList;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+
+/**
+ * Generates a .js (javascript) file in the reports directory containing the apps and their dependencies.
+ */
+@RuleMetadata(phase = ReportRenderingPhase.class)
+public class CreateDependencyGraphDataRuleProvider extends AbstractRuleProvider
+{
+
+   private static final String APP_DEPENDENCY_GRAPH_JS_FILENAME = "app_dependencies_graph.js";
+   private static final String JS_DATA_FUNCTION_NAME = "app_dependencies";
+
+   @Override
+   public Configuration getConfiguration(RuleLoaderContext ruleLoaderContext)
+   {
+      return ConfigurationBuilder.begin()
+               .addRule()
+               .when(SourceMode.isDisabled())
+               .perform(new GraphOperation()
+               {
+                  @Override
+                  public void perform(GraphRewrite event, EvaluationContext context)
+                  {
+                     generateData(event);
+                  }
+               });
+   }
+
+   private void generateData(GraphRewrite event)
+   {
+      final GraphContext graphContext = event.getGraphContext();
+      final ReportService reportService = new ReportService(graphContext);
+      final List<DependenciesReportModel> dependenciesReportModels = graphContext.service(DependenciesReportModel.class)
+               .findAll();
+
+      dependenciesReportModels.stream().forEach(dependenciesReportModel -> {
+         ProjectModel application = dependenciesReportModel.getProjectModel();
+
+         // in case of shared libraries ("virtual" project) we don not
+         // generate the dependency graph
+         if (application != null && ProjectModel.TYPE_VIRTUAL.equals(application.getProjectType()))
+            return;
+
+         Path dataDirectory = reportService.getReportDataDirectory();
+         List<DependencyReportDependencyGroupModel> dependencyReportDependencyGroupModels = dependenciesReportModel
+                  .getArchiveGroups();
+         Map<String, DependencyGraphItem> items = new HashMap<>(dependencyReportDependencyGroupModels.size() + 1);
```

Comment:
What's the +1 for?

Reply:
It's for the analyzed application itself that is not part of the its own `List<DependencyReportDependencyGroupModel> dependencyReportDependencyGroupModels`